dev/io.openliberty.microprofile.telemetry.internal.common/src/io/openliberty/microprofile/telemetry/internal/common/OpenTelemetryInfoFactoryImpl.java

```diff
@@ -0,0 +1,329 @@
+            }
+        } catch (Exception e) {
+            Tr.warning(tc, "applicationStopped", "failed to dispose of OpenTelemetry");//TODO better message
+        }
+
+    }
+
+    //Adds the service name to the resource attributes
+    private static Resource customizeResource(Resource resource, ConfigProperties c) {
+        ResourceBuilder builder = resource.toBuilder();
+        builder.put(ResourceAttributes.SERVICE_NAME, getServiceName(c));
+        return builder.build();
+    }
+
+    //Uses application name if the user has not given configured service.name resource attribute
+    private static String getServiceName(ConfigProperties c) {
+        String appName = c.getString(SERVICE_NAME_PROPERTY);
+        ComponentMetaData cmd = ComponentMetaDataAccessorImpl.getComponentMetaDataAccessor().getComponentMetaData();
+        if (appName == null) {
+            if (cmd != null) {
+                appName = cmd.getModuleMetaData().getApplicationMetaData().getName();
+            }
+        }
+        return appName;
+    }
+
+    //Interfaces and private classes only relevent to this factory.
+
+    public interface OpenTelemetrySdkBuilderSupplier {
+        public AutoConfiguredOpenTelemetrySdkBuilder getPartiallyConfiguredOpenTelemetrySDKBuilder();
+    }
+
+    /*
+     * There are two race conditions we need to protect against.
+     *
+     * On app startup/shutdown we need to swap the holder associated with the application.
+     * We protect against race conditions here by using an AtomicReference.
+     *
+     * Within the context of an application's lifecycle we need to ensure OpenTelemetryInfo
+     * is only created once. OpenTelemetryInfoReference handles this.
+     */
+    private class OpenTelemetryInfoReference extends AtomicReference<OpenTelemetryInfoWrappedSupplier> {
+
+        private static final long serialVersionUID = -4884222080590544495L;
+    }
+
+    private class OpenTelemetryInfoWrappedSupplier {
+        private Supplier<OpenTelemetryInfo> supplier;
+        private volatile boolean hasRun = false;
+        private OpenTelemetryInfo openTelemetryInfo = null;
+
+        public OpenTelemetryInfoWrappedSupplier(Supplier<OpenTelemetryInfo> supplier) {
+            this.supplier = supplier;
+        }
+
+        public OpenTelemetryInfo get() throws InterruptedException, ExecutionException {
+            if (openTelemetryInfo != null) {
+                return openTelemetryInfo;
+            }
+
+            synchronized (this) {
+                hasRun = true;
+                if (openTelemetryInfo != null) {
+                    return openTelemetryInfo;
+                }
+
+                openTelemetryInfo = supplier.get();
+                return openTelemetryInfo;
+            }
+        }
+
+        public boolean lock() {
+            synchronized (this) {
+                supplier = OpenTelemetryInfoFactoryImpl::createDisposedOpenTelemetryInfo;
+                return hasRun;
+            }
+        }
```

Comment:
I think this would be simpler if this class handled disposal as well.

Reply:
Check